fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java

```diff
@@ -1375,26 +1375,51 @@ private void createAnulacionNoveltyNews(Loan loan, LocalDate writeOffDate) {
 
     private void createNoveltyNews(Loan loan, LocalDate transactionDate, InsuranceIncidentType incidentType) {
         // Fetch the Insurance Incident based on the provided type
+        InsuranceIncident incident = fetchValidIncident(incidentType);
+
+        if (incidentType == InsuranceIncidentType.DEFINITIVE_RESTRUCTURING_CANCELLATION) {
+            handleDefinitiveRestructuringCancellation(loan, incident, transactionDate);
+        } else {
+            processLoanChargesForNoveltyNews(loan, incident, transactionDate);
+        }
+    }
+
+    private InsuranceIncident fetchValidIncident(InsuranceIncidentType incidentType) {
         InsuranceIncident incident = this.insuranceIncidentRepository.findByIncidentType(incidentType);
-        if (incident == null || (!incident.isValid())) {
+        if (incident == null || !incident.isValid()) {
             throw new InsuranceIncidentNotFoundException(incidentType.name());
         }
+        return incident;
+    }
 
-        // Iterate through loan charges and create novelty news as needed
-        for (LoanCharge loanCharge : loan.getCharges()) {
-            if (loanCharge.getAmountOutstanding(loan.getCurrency()).isGreaterThanZero()) {
-                if ((incident.isMandatory() && loanCharge.isMandatoryInsurance())
-                        || (incident.isVoluntary() && loanCharge.isVoluntaryInsurance())) {
-                    BigDecimal cumulative = BigDecimal.ZERO;
-                    InsuranceIncidentNoveltyNews insuranceIncidentNoveltyNews = InsuranceIncidentNoveltyNews.instance(loan, loanCharge,
-                            null, incident, transactionDate, cumulative);
+    private void handleDefinitiveRestructuringCancellation(Loan loan, InsuranceIncident incident, LocalDate transactionDate) {
+        boolean doesIncidentExist = this.insuranceIncidentNoveltyNewsRepository.existsByLoanAndIncident(loan.getId(), incident.getId());
+        if (!doesIncidentExist) {
+            LoanCharge loanCharge = loan.getLoanCharges().stream()
```

Comment:
@oluexpert99 The purpose of below if condition is to make sure novelty news is created for either voluntary or mandatory charges or both depending upon the selection of charge types for the novelty news. Here you have restricted the code to create novelty news for only mandatory insurance even if the loan has voluntary insurance and voluntary insurance is selected for the novelty incident

Reply:
@oluexpert99 existsByLoanAndIncident method should check if incident is created against a particular charge or not if needed

Reply:
Hi @faheem205  , This check is just for DEFINITIVE_RESTRUCTURING_CANCELLATION
The requirement is that the event be shown just once for loan .

Reply:
@oluexpert99 So if a loan has only voluntary insurance then this incident will not be recorded?

Reply:
@faheem205 ,Now I get what you are implying. Thanks for the catch . PR updated

Reply:
@oluexpert99 Now if loan does not have a mandatory or voluntary insurance charge then the code will try to create an incident with null loancharge and probably throw an exception

Reply:
fixed